### PR TITLE
feat(update-user-password):user updating password

### PR DESCRIPTION
**Pull request by Ruthuwamahoro**
-users can update their password when they have validly registered -users must provide their old password, new password and confirm their password

issue in referencing?

Questions:

What does this PR do?

Password Change Functionality Implementation

Description of Task to be completed?
- Ensuring that the user attempting to change the password is a valid registered user with a validated account
- Requires the user to provide their current password for authentication purposes, along with the new password they want to set  and confirmation to avoid errors due to mist, typed password

-For security reasons, ensure the new password is securely hashed before updating it in the database.
- Prevents unauthorised password changes and ensures that the user can only change their password if they correctly provide their old password and confirm the new password.

#### How should this be manually tested?

-run "npm run install"
-run "npm run migrate"
-run "npm run seed"
-run "npm run dev"
 

 background context you want to provide?

Not applicable

#### What are the relevant pivotal tracker/Trello stories?

- id of the project
  #187300151

#### Questions:

-Not applicable

### CODE REVIEW

File: src/controllers/changePassword.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+      });
+      logger.info(userId);
+      if (!userId) return done(null, false);
+      return done(null, userId);
+    } catch (err) {
+      logger.info(err);
+      return done(err, false);
+    }
+  })
+);
```
Comment (fab-ryan) [inline src/controllers/changePassword.ts]: Please create strategy to different file

File: src/controllers/changePassword.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+              .join(', ');
+            return res.status(400).json({ status: 400, error: messageError });
+          }
+          logger.info('----------------------', checkErrors);
+          const { oldPassword, newPassword } = req.body;
+
+          const checkOldPassword = await bcrypt.compare(
+            oldPassword,
+            user.password
+          );
```
Comment (fab-ryan) [inline src/controllers/changePassword.ts]: use validation as middlware not like this

**Review by fab-ryan — CHANGES_REQUESTED** [review]
first rebase to develop branch

File: src/controllers/loginController.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    async (err: any, user: any) => {
+      // test if uer not found
+
+      if (!user) return sendResponse<null>(res, 400, null, 'User not found');
+
+      // generate token
+
+      const { secret } = config();
+      const token = jwt.sign({ id: user.id }, secret, { expiresIn: '2h' });
+
```
Comment (fab-ryan) [inline src/controllers/loginController.ts]: I told you to separate this with current codes

File: src/controllers/changePasswordController.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  passport.authenticate(
+    'jwt',
+    { session: false },
+    async (err: Error | null, user: User) => {
+      try {
+        // if (err) {
+        //   return sendResponse<null>(res, 400, null, 'user not found');
+        // }
+        if (user) {
+          const { oldPassword, newPassword } = req.body;
```
Comment (fab-ryan) [inline src/controllers/changePasswordController.ts]: same here remove those stategy put it in app.ts and first create it's files you use it in global

**Review by fab-ryan — COMMENTED** [review]
please optimize your codes

File: src/controllers/changePasswordController.ts (hunk 4 above)
Comment (Ruthuwamahoro) [reply src/controllers/changePasswordController.ts]: I have created generalStrategy.ts in helper folder and I made changes.

File: src/controllers/loginController.ts (hunk 3 above)
Comment (Ruthuwamahoro) [reply src/controllers/loginController.ts]: The only way I did this is to remove that strategy and put it in app.ts.Could you explain further?

**Review by musabehonore — APPROVED** [review]

File: src/controllers/changePasswordController.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  password: string;
+}
+/// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+const updatePassword = async (
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  return passport.authenticate(
```
Comment (fab-ryan) [inline src/controllers/changePasswordController.ts]: this are handled in middlwares

**Review by RWEMAREMY — APPROVED** [review]
good job

File: src/middleware/validatePassword.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  (schema: Joi.ObjectSchema, type: requestType = requestType.body) =>
+  (req: Request, res: Response, next: NextFunction) => {
+    try {
+      const { error } = schema.validate(req.body, { abortEarly: false });
+
+      if (error) {
+        const errors: errorInterface[] = [];
+        error.details.forEach(err => {
+          const field = err.path[0] as string;
+          if (field === 'password') {
```
Comment (fab-ryan) [inline src/middleware/validatePassword.ts]: this file is not necessary we already have validationn middleware use that

File: src/routers/userRoute.ts [hunk 7]
```diff
@@ -16,5 +19,10 @@ router.post(
   validationMiddleware(loginUserSchema),
   validatingUser
 );
+router.post(
```
Comment (fab-ryan) [inline src/routers/userRoute.ts:22]: use validation middlware only pass schemas

File: test/updatePassword.test.ts [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+      .post('/api/users/password')
+      .set('Authorization', `Bearer ${Token}`)
+      .send(userPassword)
+      .end((err, res) => {
+        expect(res).to.have.status(200);
+        expect(res.body).to.have.property('message');
+        expect(res.body.message).to.equal('change Password Successfully');
+      });
+  });
+  it('when old password is not the same as the one stored in database', async () => {
```
Comment (fab-ryan) [inline test/updatePassword.test.ts]: no test that trigger 500

**Review by fab-ryan — CHANGES_REQUESTED** [review]
Work on the requests

**Review by manzidavidrw — APPROVED** [review]

File: src/middleware/authentication.ts [hunk 9]
```diff
@@ -11,14 +11,15 @@ export const isAuthenticated = (
...
-   (err: Error, user: UserAttributes) => {
+    (err: Error, user: UserAttributes) => {
       if (!user) {
         return sendResponse(res, 403, null, 'Please login...');
       }
       const currUser = {
         id: user.id,
         name: user.name,
         email: user.email,
+        password: user.password,
```
Comment (fab-ryan) [inline src/middleware/authentication.ts]: we don't need password please remove it  you have id of user please remove password

File: src/routers/userRoute.ts [hunk 10]
```diff
@@ -16,5 +19,11 @@ router.post(
   validationMiddleware(loginUserSchema),
   validatingUser
```
Comment (fab-ryan) [inline src/routers/userRoute.ts]: please cchange the name of this controller to loginController

**Review by fab-ryan — CHANGES_REQUESTED** [review]
Please correct those mistakes

**Review by Mukunzijames — APPROVED** [review]

**Review by fab-ryan — APPROVED** [review]

**Pull request merged**